proposals/NNNN-unlock-existential-types-for-all-protocols.md

```diff
@@ -0,0 +1,405 @@
+
+```swift
+protocol Equatable {
+
+  static func == (lhs: Self, rhs: Self) -> Bool
+  //                   ^~~~       ^~~~
+  // Use of 'Self' in a non-covariant position.
+
+}
+
+struct Dog : Equatable { ... }
+
+
+let ownerName = "Paul" as Equatable ✅
+
+let petName = "Alex" as Equatable ✅
+
+let dog = Dog(named: petName) as Equatable ✅
+
+
+ownerName == petName ❌
+// Error: TBD
+
+ownerName == dog ❌
+// Error: TBD
+```
+
+> **Rationale**: In the above case, it would be reasonable to assume that `ownerName` and `petName` could be compared, since they're both bound to `String`. However, that would be inaccurate, as they're actually bound to the existential type of `Equatable`. In spite of this, it could, still, be argued that `Equatable`'s `==(lhs:rhs:)` operator could unwrap its boxed `Self` parameters to the unboxed type ( `String` in this case), and thus compare its two `Self` parameters. There are, though, fundental problems with this approach, the most important of which being the fact that two `Self` parameters could be bound to different types, which would introduce undefined behavior. The example of comparing `ownerName` and `dog`, demonstrates this perfectly, as – besides being unsound – it poses the question of how it should be handled, if it were valid.
+
+### Types of Requirements
+
+As for the type that a requirement will be bound to when accessed from an existential type, it should be noted that all requirements that do _not_ reference `Self` or any associated type _will_ retain their type as is. However, requirements that do _not_ meet these criteria will be bound to an existential type of `Self` or the relevant associated type accordingly. For instance:
+
+```swift
+
+protocol User { 
+
+  associatedtype ID : Hashable 
+
+
+  var id: ID { get }
+
+  var username: String { get } 
+
+}
+
+struct RegularUser : User { ... }
+
+
+let regularUser = RegularUser(id: UUID(), username: "i❤️Swift")
+
+let userExistential = regularUser as User
+
+
+userExistential.username ℹ️
+// Output: i❤️Swift
+
+type(of: userExistential.username) ℹ️
+// Output: String
+
+
+type(of: userExistential.id) ℹ️
+// Output: Hashable
+```
+
+Looking at the above, one will notice that the `id` requirement of `User` is available despite being bound to the associated type `ID`; thus, the compiler exposes the existential type of `Hashable` as `id`'s type – since, the associated type is _unknown_ in the eyes of `userExistential`. Of course, `username` – whose type is known to be `String` – retains its type, as seen in the above example.
+
+
+## Source compatibility
+
+This is an additive change with _no_ impact on **source compatibility**.
+
+
+## Effect on ABI stability
+
+This is an additive change with _no_ impact on **ABI stability**.
+
+
+## Effect on API resilience
+
+This is an additive change with _no_ impact on **API resilience**.
```

Comment:
We should mention that adding a defaulted associated type becomes both a binary- and source-compatible change.